[AllocBoxToStack] Don't destroy in dead-ends.

nate-chandler opened the pull request:
It is valid to leak a value on paths into dead-end regions. Specifically, it is valid to leak an `alloc_box`.  Thus, "final releases" in dead-end regions may not destroy the box and consequently may not release its contents.  Therefore it's invalid to lower such final releases to `dealloc_stack`s, let alone `destroy_addr`s.  The in-general invalidity of that transformation results in miscompiling whenever a box is leaked and its projected address is used after such final releases.

Fix this by not treating final releases as boundary markers of the `alloc_box` and not lowering them to `destroy_addr`s and `dealloc_stack`s.

rdar://158149082

nate-chandler committed:
[AllocBoxToStack] Restore isDeadEnd check.

The rewrite was missing the intentional omission of `dealloc_stack`s
corresponding to `[dead_end]` `dealloc_box`es.  Add the necessary
bridging to get to parity with the original.

Without this check, `dealloc_box [dead_end]`s are promoted to
`dealloc_stack`s but the memory projected out of such `alloc_box`s need
not be valid.

rdar://159271158

nate-chandler committed:
[AllocBoxToStack] Don't destroy in dead-ends.

It is valid to leak a value on paths into dead-end regions.
Specifically, it is valid to leak an `alloc_box`.  Thus, "final
releases" in dead-end regions may not destroy the box and consequently
may not release its contents.  Therefore it's invalid to lower such final
releases to `dealloc_stack`s, let alone `destroy_addr`s.  The in-general
invalidity of that transformation results in miscompiling whenever a box
is leaked and its projected address is used after such final releases.

Fix this by not treating final releases as boundary markers of the
`alloc_box` and not lowering them to `destroy_addr`s and
`dealloc_stack`s.

rdar://158149082

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocBoxToStack.swift:
```diff
@@ -359,12 +359,25 @@ private func createAllocStack(for allocBox: AllocBoxInst, flags: Flags, _ contex
     let loc = allocBox.location.asCleanup.withScope(of: destroy.location)
     Builder.insert(after: destroy, location: loc, context) { builder in
-      if  !unboxedType.isTrivial(in: allocBox.parentFunction), !(destroy is DeallocBoxInst) {
+      if deadEndBlocks.isDeadEnd(destroy.parentBlock) {
```
This is not correct for the corner case the alloc_box is within an infinite loop, e.g.
```
bb1:
  %1 = alloc_box
  destroy_value %1
  br bb1
```
If we are not inserting the destroy_addr we will leak the value. 

I don't have a good idea how to handle this without replicating half of the ossa-lifetime-completion code. Maybe a simple solution is to just not promote alloc_box in dead-end infinite loops.

Anyway, this problem should be solved once we have complete lifetimes and OSSA throughout the pipeline.
Maybe also worth adding a TODO to remove this check once we have complete lifetimes

eeckstein commented on lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:
```diff
@@ -693,9 +695,19 @@ static bool rewriteAllocBoxAsAllocStack(AllocBoxInst *ABI) {
                          ABI->getBoxType(), ABI->getModule().Types, 0));
   auto Loc = CleanupLocation(ABI->getLoc());
 
+  auto *deb = deba.get(ABI->getFunction());
```
No need to fix this in the old pass. As long as we can successfully build the stdlib with this and the bootstrapping CI job does not fail, we are fine.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler committed:
[NFC] OptUtils: Promote isInLoop to internal.

It will be used by both AllocBoxToStack in addition to StackPromotion
and for the same reason.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler committed:
[MemoryLifetimeVerifier] Permit leaks in dead-ends

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test macos platform

eeckstein reviewed (approved):
lgtm!
